ibis/backends/clickhouse/__init__.py

```diff
@@ -140,10 +259,86 @@ def list_tables(self, like=None, database=None):
         databases = list(data[0])
         return self._filter_with_like(databases, like)
 
+    def _apply_limit(self, sql, limit):
+        if limit == "default":
+            limit = ibis.options.sql.default_limit
+
+        if limit is not None:
+            try:
+                limexpr = sql.args["limit"]
+            except KeyError:
+                try:
+                    sql = sql.limit(limit)
+                except AttributeError:
+                    sql = sg.select("*").from_(sql).limit(limit)
+            else:
+                curlimit = int(limexpr.args["expression"].args["this"])
+                newlimit = min(limit, curlimit)
+                sql.args["limit"] = limexpr.replace(
+                    sg.exp.Limit(expression=sg.exp.Literal.number(newlimit))
+                )
+        return sql
+
+    def execute(
+        self,
+        expr: ir.Expr,
+        limit: str | None = 'default',
+        external_tables: Mapping[str, pd.DataFrame] | None = None,
+        **kwargs: Any,
+    ) -> Any:
+        """Execute an expression."""
+        table_expr = expr.as_table()
+        sql = self.compile(table_expr, limit=limit, **kwargs)
+        self._log(sql)
+
+        for memtable in graph.traverse(util.find_memtables, expr.op()):
```

Comment:
`finf_memtables` looks like an analysis function, could we move it there?

Reply:
Moved!